cmd/influxd/node.go

```diff
@@ -0,0 +1,569 @@
+		}
+	}
+
+	if n.Broker != nil {
+		if err := n.Broker.Close(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// creates and initializes a server.
+func (n *Node) openServer(joinURLs []url.URL) *influxdb.Server {
+
+	// Create messaging client to the brokers.
+	c := influxdb.NewMessagingClient(*n.ClusterURL())
+	c.SetURLs(joinURLs)
+
+	if err := c.Open(filepath.Join(n.config.Data.Dir, messagingClientFile)); err != nil {
+		log.Fatalf("messaging client error: %s", err)
+	}
+
+	// If no URLs exist on the client the return an error since we cannot reach a broker.
+	if len(c.URLs()) == 0 {
+		log.Fatal("messaging client has no broker URLs")
+	}
+
+	// Create and open the server.
+	s := influxdb.NewServer()
+
+	s.WriteTrace = n.config.Logging.WriteTracing
+	s.RetentionAutoCreate = n.config.Data.RetentionAutoCreate
+	s.RecomputePreviousN = n.config.ContinuousQuery.RecomputePreviousN
+	s.RecomputeNoOlderThan = time.Duration(n.config.ContinuousQuery.RecomputeNoOlderThan)
+	s.ComputeRunsPerInterval = n.config.ContinuousQuery.ComputeRunsPerInterval
+	s.ComputeNoMoreThan = time.Duration(n.config.ContinuousQuery.ComputeNoMoreThan)
+	s.Version = version
+	s.CommitHash = commit
+
+	// Open server with data directory and broker client.
+	if err := s.Open(n.config.Data.Dir, c); err != nil {
+		log.Fatalf("failed to open data node: %v", err.Error())
+	}
+	log.Printf("data node(%d) opened at %s", s.ID(), n.config.Data.Dir)
+
+	// Give brokers time to elect a leader if entire cluster is being restarted.
+	time.Sleep(1 * time.Second)
+
+	if s.ID() == 0 {
+		n.joinOrInitializeServer(s, *n.ClusterURL(), joinURLs)
+	} else {
+		log.Printf("data node already member of cluster. Using existing state and ignoring join URLs")
+	}
+
+	return s
+}
+
+// joinOrInitializeServer joins a new server to an existing cluster or initializes it as the first
+// member of the cluster
+func (n *Node) joinOrInitializeServer(s *influxdb.Server, u url.URL, joinURLs []url.URL) {
+	// Create data node on an existing data node.
+	for _, joinURL := range joinURLs {
+		if err := s.Join(&u, &joinURL); err == influxdb.ErrDataNodeNotFound {
+			// No data nodes could be found to join.  We're the first.
+			if err := s.Initialize(u); err != nil {
+				log.Fatalf("server initialization error(1): %s", err)
+			}
+			log.Printf("initialized data node: %s\n", (&u).String())
+			return
+		} else if err != nil {
+			// does not return so that the next joinURL can be tried
+			log.Printf("join: failed to connect data node: %s: %s", (&u).String(), err)
+		} else {
+			log.Printf("join: connected data node to %s", u)
+			return
+		}
+	}
+
+	if len(joinURLs) == 0 {
+		if err := s.Initialize(u); err != nil {
+			log.Fatalf("server initialization error(2): %s", err)
+		}
+		log.Printf("initialized data node: %s\n", (&u).String())
+		return
+	}
+
+	log.Fatalf("join: failed to connect data node to any specified server")
+}
+
+// Leave removes the node from the cluster
+func (n *Node) Leave() error {
```

Comment:
I'm not sure we can return an error here. If we are leaving, we don't know what data nodes succeeded or failed.  Other data nodes could get the message for a remove server, remove all reference to them (success), and the server being removed could fail and then still be running and shouldn't be.